test/lib/pwa.js

```diff
@@ -32,39 +36,114 @@ var assert = require('chai').assert;
+      simpleMock.mock(libPwa, 'db').returnWith(db);
+      return libPwa.updateLighthouseInfo(pwa).should.be.fulfilled.then(updatedPwa => {
+        assert.equal(libLighthouse.fetchAndSave.callCount, 1);
+        assert.equal(libLighthouse.fetchAndSave.lastCall.args[0], '123456789');
+        assert.equal(db.update.callCount, 1);
+        assert.equal(updatedPwa.lighthouseScore, 82);
+      });
+    });
+  });
 
-          // Mock libImages and bd to avoid making real calls
-          simpleMock.mock(libImages, 'fetchAndSave').resolveWith(manifest.getBestIconUrl());
-          pwaLib.libImages = libImages;
-          simpleMock.mock(db, 'update').returnWith(pwa);
-          pwaLib.db = db;
+  describe('#save (core logic)', () => {
+    afterEach(() => {
+      simpleMock.restore();
+    });
+    it('performs all the save steps', () => {
+      // Mock findByManifestUrl and bd to avoid making real calls
+      simpleMock.mock(libPwa, 'findByManifestUrl').resolveWith(pwa);
+      simpleMock.mock(libManifest, 'fetchManifest').resolveWith(manifest);
+      simpleMock.mock(libPwa, 'fetchMetadataDescription').resolveWith('test-description');
+      simpleMock.mock(db, 'update').returnWith(pwa);
+      simpleMock.mock(libPwa, 'updateIcon').resolveWith(pwa);
+      simpleMock.mock(libPwa, 'updateLighthouseInfo').resolveWith(pwa);
+      simpleMock.mock(libPwa, 'libManifest').returnWith(libManifest);
+      simpleMock.mock(libPwa, 'db').returnWith(db);
 
-          var promiseUpdateIcon = pwaLib.updateIcon(pwa, manifest);
-          assert.equal(libImages.fetchAndSave.callCount, 1);
-          assert.equal(libImages.fetchAndSave.lastCall.args[0], 'https://s1.trrsf.com/fe/zaz-morph/_img/launcher-icon.png?v2');
-          assert.equal(libImages.fetchAndSave.lastCall.args[1], '123456789.png');
-          return promiseUpdateIcon.should.eventually.have.property('iconUrl');
-        });
+      return libPwa._save(pwa).should.be.fulfilled.then(updatedPwa => {
+        assert.equal(libPwa.findByManifestUrl.callCount, 1);
+        assert.equal(libPwa.findByManifestUrl.lastCall.args[0], MANIFEST_URL);
+        assert.equal(libManifest.fetchManifest.callCount, 1);
+        assert.equal(libManifest.fetchManifest.lastCall.args[0], MANIFEST_URL);
+        assert.equal(updatedPwa.manifest, manifest);
+        assert.equal(libPwa.fetchMetadataDescription.callCount, 1);
+        assert.equal(libPwa.fetchMetadataDescription.lastCall.args[0], pwa);
+        assert.equal(updatedPwa.metaDescription, 'test-description');
+        assert.equal(db.update.callCount, 1);
+        assert.equal(libPwa.updateIcon.callCount, 1);
+        assert.equal(libPwa.updateLighthouseInfo.callCount, 1);
+      });
+    });
+    it('handles E_MANIFEST_ERROR error', () => {
+      // Mock findByManifestUrl to avoid making real calls
+      simpleMock.mock(libPwa, 'findByManifestUrl').rejectWith(new Error('Testing error'));
```

Comment:
This results in an ugly stack trace in the middle of the mocha output, is there any way to get rid of this?

```text
Updated PWA Lighthouse info for:  123456789
      ✓ sets lighthouseScore
    #save (core logic)
      ✓ performs all the save steps
Error: Testing error
    at Context.it (/Users/stillers/workspace/gulliver/test/lib/pwa.js:133:63)
    at callFn (/Users/stillers/workspace/gulliver/node_modules/mocha/lib/runnable.js:334:21)
    at Test.Runnable.run (/Users/stillers/workspace/gulliver/node_modules/mocha/lib/runnable.js:327:7)
    at Runner.runTest (/Users/stillers/workspace/gulliver/node_modules/mocha/lib/runner.js:429:10)
    at /Users/stillers/workspace/gulliver/node_modules/mocha/lib/runner.js:535:12
    at next (/Users/stillers/workspace/gulliver/node_modules/mocha/lib/runner.js:349:14)
    at /Users/stillers/workspace/gulliver/node_modules/mocha/lib/runner.js:359:7
    at next (/Users/stillers/workspace/gulliver/node_modules/mocha/lib/runner.js:285:14)
    at Immediate.<anonymous> (/Users/stillers/workspace/gulliver/node_modules/mocha/lib/runner.js:327:5)
    at runCallback (timers.js:574:20)
    at tryOnImmediate (timers.js:554:5)
    at processImmediate [as _immediateCallback] (timers.js:533:5)
      ✓ handles E_MANIFEST_ERROR error
    #save (validation logic)
      ✓ rejects on null pwa
      ✓ rejects if not passed a Pwa object
      ✓ rejects if passed a Pwa object without a manifestUrl
      ✓ rejects if passed a Pwa object with an invalid manifestUrl
      ✓ rejects if passed a Pwa object with an invalid user.id
      ✓ fulfills if passed a valid Pwa objectid
```